graphqlapi/src/main/java/io/stargate/graphql/graphqlservlet/HTTPAwareContextImpl.java

```diff
@@ -97,4 +111,61 @@ private boolean hasJWTToken() {
   public Optional<DataLoaderRegistry> getDataLoaderRegistry() {
     return Optional.ofNullable(dataLoaderRegistry);
   }
+
+  public BatchContext getBatchContext() {
+    return batchContext;
+  }
+
+  /**
+   * Encapsulates logic to add multiple statements contained in the same operation that need to be
+   * executed in a batch.
+   */
+  public static class BatchContext {
+    private final List<String> statements = new ArrayList<>();
+    private final Object statementsLock = new Object();
+    private final CompletableFuture<ResultSet> executionFuture = new CompletableFuture<>();
+    private volatile DataStore dataStore;
+    private final AtomicInteger dataStoreCounter = new AtomicInteger();
+
+    public CompletableFuture<ResultSet> getExecutionFuture() {
+      return executionFuture;
+    }
+
+    public List<String> getStatements() {
+      synchronized (statementsLock) {
+        return statements;
+      }
+    }
+
+    public void setExecutionResult(CompletableFuture<ResultSet> result) {
+      result.whenComplete(
+          (r, e) -> {
+            if (e != null) {
+              executionFuture.completeExceptionally(e);
+            } else {
+              executionFuture.complete(r);
+            }
+          });
+    }
+
+    public void setExecutionResult(Exception ex) {
+      executionFuture.completeExceptionally(ex);
+    }
+
+    public int add(String query) {
+      synchronized (statementsLock) {
+        statements.add(query);
+        return statements.size();
+      }
+    }
+
+    public int setDataStore(DataStore dataStore) {
+      this.dataStore = dataStore;
+      return dataStoreCounter.incrementAndGet();
```

Comment:
This is not likely to be a bottleneck and we're already using synchronization. What about just making this a synchronized method? Then we can remove `volatile` and the atomic counter.

```java 
   // This could return a boolean instead of throwing, but this is an exceptional case and a return value can be accidentally ignored.
    public synchronized void setDataStore(DataStore dataStore) {
      if (this.dataStore != null)
        throw new IllegalStateException("Datastore already set");
      this.dataStore = dataStore;
    }

    public synchronized Optional<DataStore> getDataStore() {
      return Optional.ofNullable(this.dataStore);
    }
```

Reply:
I've used a single atomic reference for both.